### PR TITLE
Somil55/multisource

### PR DESCRIPTION
**Pull request by somil55**
This function can be used to find shortest path from a subset of vertices to all vertices.This function by default considers all vertices unless provided

### CODE REVIEW

**Comment by sbromberger** [comment]
I like the code. I need to think about the name, though :) Bikeshedding time!

File: src/shortestpaths/dijkstra.jl [hunk 1]
```diff
@@ -116,3 +116,45 @@ end
 
 dijkstra_shortest_paths(g::AbstractGraph, src::Integer, distmx::AbstractMatrix = DefaultDistance(); allpaths=false, trackvertices=false) =
 dijkstra_shortest_paths(g, [src;], distmx; allpaths=allpaths, trackvertices=trackvertices)
+
+struct MultipleDijkstraState{T, U<:Integer}<:AbstractPathState
```
Comment (sbromberger) [inline src/shortestpaths/dijkstra.jl]: Docstring for this, detailing why it's different from DijkstraState?

File: src/shortestpaths/dijkstra.jl [hunk 2]
```diff
@@ -116,3 +116,45 @@ end
...
+    dists::Matrix{T}
+    parents::Matrix{U}
+end
+
+function multisource_dijkstra_shortest_paths(
```
Comment (sbromberger) [inline src/shortestpaths/dijkstra.jl]: docstring here, please

File: src/shortestpaths/dijkstra.jl [hunk 3]
```diff
@@ -116,3 +116,45 @@ end
...
+    g::AbstractGraph,
+    sources::Vector{U} = collect(vertices(g)),
```
Comment (sbromberger) [inline src/shortestpaths/dijkstra.jl]: Why collect the vertices? This could allocate a lot of memory.

File: src/shortestpaths/dijkstra.jl [hunk 4]
```diff
@@ -116,3 +116,45 @@ end
...
+    distmx::AbstractMatrix{T} = DefaultDistance();
+    parallel=false,
```
Comment (sbromberger) [inline src/shortestpaths/dijkstra.jl]: Let's make this `parallel_multisource_dijkstra_shortest_paths` and get rid of the `parallel` argument for now.

File: test/shortestpaths/dijkstra.jl [hunk 5]
```diff
@@ -100,4 +100,171 @@
...
+        end
+      end
+    end
+
+
+    #Testing multisource On directed Graph
+    g3 = PathDiGraph(5)
+    d = float([ 0 1 2 3 4; 5 0 6 7 8; 9 10 0 11 12; 13 14 15 0 16; 17 18 19 20 0])
+
+    for g in testgraphs(g3)
```
Comment (sbromberger) [inline test/shortestpaths/dijkstra.jl]: This should be `testdigraphs()`?

File: test/shortestpaths/dijkstra.jl [hunk 6]
```diff
@@ -100,4 +100,171 @@
...
+      z  = @inferred(floyd_warshall_shortest_paths(g, d))
+      zm = @inferred(multisource_dijkstra_shortest_paths(g,d))
+      @test all(isapprox(z.dists,zm.dists))
+
```
Comment (sbromberger) [inline test/shortestpaths/dijkstra.jl]: General comment - please put spaces after commas in function calls:
```
zm = @inferred(multisource_dijkstra_shortest_paths(g, d))
@test all(isapprox(z.dists, zm.dists))
...
````

File: test/shortestpaths/dijkstra.jl [hunk 7]
```diff
@@ -100,4 +100,171 @@
...
+
+    #Testing multisource On directed Graph
+    g3 = PathDiGraph(5)
+    d = float([ 0 1 2 3 4; 5 0 6 7 8; 9 10 0 11 12; 13 14 15 0 16; 17 18 19 20 0])
+
+    for g in testdigraphs(g3)
+      z  = @inferred(floyd_warshall_shortest_paths(g, d))
+      zm = @inferred(multisource_dijkstra_shortest_paths(g, collect(1:5), d))
+      @test all(isapprox(z.dists,zm.dists))
```
Comment (sbromberger) [inline test/shortestpaths/dijkstra.jl]: General comment: please separate arguments to function calls with a comma followed by a space:
```
@test all(isapprox(z.dists, zm.dists))
```

File: test/shortestpaths/dijkstra.jl [hunk 8]
```diff
@@ -100,4 +100,171 @@
...
+          else
+            @test length(state.predecessors[j]) == 0
+          end
+        end
+      end
+    end
+
+
+
+    #Testing on 100 vertices and 500 vertices random graphs
```
Comment (sbromberger) [inline test/shortestpaths/dijkstra.jl]: Too big. use a `(10, 50)`. We don't need huge loops in tests. Also, add self loops explicitly if you want them.

Comment (somil55) [inline test/shortestpaths/dijkstra.jl]: ok....working on this

**Comment by sbromberger** [comment]
Erm. this needs a rebase.

File: src/shortestpaths/dijkstra.jl [hunk 9]
```diff
@@ -116,3 +116,72 @@ end
...
+struct MultipleDijkstraState{T, U<:Integer}<:AbstractPathState
+    dists::Matrix{T}
+    parents::Matrix{U}
+end
+
+@doc_str """
+    multisource_dijkstra_shortest_paths(g, sources=vertices(g), distmx=DefaultDistance())
+    parallel_multisource_dijkstra_shortest_paths(g, sources=vertices(g), distmx=DefaultDistance())
+
+This function computes the shortest paths between all pairs of vertices in graph `g` by running
```
Comment (sbromberger) [inline src/shortestpaths/dijkstra.jl]: The docstring should be active voice:
"Compute the shortest paths .... Return ____"

File: src/shortestpaths/dijkstra.jl [hunk 10]
```diff
@@ -116,3 +116,72 @@ end
...
+[`dijkstra_shortest_paths`] for every vertex and using an optional list of source vertex `sources` and
+an optional distance matrix `distmx`. Return a [`MultipleDijkstraState`](@ref) with relevant
+traversal information.
+"""
+function multisource_dijkstra_shortest_paths{T}(
```
Comment (sbromberger) [inline src/shortestpaths/dijkstra.jl]: Do we need this? How is it better than
```
mypaths = [dijkstra_shortest_paths(g, s) for s in sources]
```
?

Comment (somil55) [inline src/shortestpaths/dijkstra.jl]: We can certainly do it the way you wrote above. But my aim for writing this function is to provide a function similar to floyd-warshall but much faster than that. Also it is a lot more easier to extract parallel performance from this as compared to floyd-warshall.

Comment (sbromberger) [inline src/shortestpaths/dijkstra.jl]: I was referring specifically to the non-parallel version of `multisource_dijkstra_shortest_paths`. Is there a benefit over `[dijkstra_shortest_paths(g, s) for s in sources]`?

Comment (somil55) [inline src/shortestpaths/dijkstra.jl]: No.....We can scrap this for sure 
[dijkstra_shortest_paths(g, s) for s in sources] this is equivalent to sequential version. Infact gives more info in same time.

Comment (sbromberger) [inline src/shortestpaths/dijkstra.jl]: Understood. Let's scrap it for now.

**Comment by sbromberger** [comment]
@jpfairbanks can you help me bikeshed the name? I'm ok leaving it as is (that is, `parallel_multisource_dijkstra_shortest_paths` but it seems... wordy.)

File: src/shortestpaths/dijkstra.jl (hunk 10 above)
Comment (somil55) [inline src/shortestpaths/dijkstra.jl]: Yea sure. I will push the changes :)

**Comment by sbromberger** [comment]
You're going to have to rebase this - if I commit this as-is we lose most of the work we've done to master in the past few days.

**Comment by somil55** [comment]
#661 opened

**Pull request closed without merging**